src/protocol/mod.rs

```diff
@@ -123,6 +124,45 @@ macro_rules! repeat64str {
     }
 }
 
+/// Bounds-checked index type for protocol steps indexed 0..64.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct StepIndex64(u8);
+
+impl StepIndex64 {
+    pub const BOUND: usize = 64;
+    const BOUND_U8: u8 = 64;
+}
+
+impl<'a> Index<StepIndex64> for [&'a str; 64] {
+    type Output = &'a str;
+
+    fn index(&self, index: StepIndex64) -> &Self::Output {
+        &self[usize::from(index.0)]
+    }
+}
+
+#[derive(Debug)]
+pub struct StepIndexFromIntError;
+
+macro_rules! step_index_try_from {
+    ($ty:ty) => {
+        impl TryFrom<$ty> for StepIndex64 {
+            type Error = StepIndexFromIntError;
+
+            fn try_from(value: $ty) -> Result<Self, Self::Error> {
+                match u8::try_from(value) {
+                    Ok(v) if v < Self::BOUND_U8 => Ok(Self(v)),
+                    _ => Err(StepIndexFromIntError),
+                }
+            }
+        }
+    };
+}
+
+step_index_try_from!(usize);
+step_index_try_from!(u32);
+step_index_try_from!(i32);
```

Comment:
it is probably unlikely that we need this except in tests?

Reply:
It was used for numeric literals (not just in tests), but that can be fixed by specifying `u32` suffix on the literals, which seems like the right tradeoff given that a negative step index is nonsensical.